### PR TITLE
Add support for Azure storage account user assigned identity

### LINKED ISSUE #5710: Add support for Azure user assigned managed identity
**Is your feature request related to a problem? Please describe.**
We can't configure Loki running in a Kubernetes cluster with a user assigned managed identity to access an Azure storage account. We use [aad-pod-identity](https://github.com/Azure/aad-pod-identity) to give services their identity in our AKS clusters, much like IRSA in EKS, and need Loki to support this.

**Describe the solution you'd like**
Allow the user to provide the user assigned identity id when configuring the Azure storage settings.

**Describe alternatives you've considered**
n/a

**Additional context**
The current Azure package versions should support this based on the Thanos implementation. @Ottovsky mentioned this in https://github.com/grafana/loki/issues/4256.

### PR DESCRIPTION
**Pull request by stevehipwell**
**What this PR does / why we need it**:

This PR adds support for authenticating to an Azure storage account via a user defined identity and is based on the Thanos implementation.

This PR also tidies up the existing code for Azure storage account and makes the documentation easier to understand.

**Which issue(s) this PR fixes**:
Fixes #5710.

**Special notes for your reviewer**:
I haven't tested this in Azure yet as I'd like a review in principal first.

**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/storage/bucket/azure/config_test.go
Comment (chaudum) [inline pkg/storage/bucket/azure/config_test.go:16]: nit: Why `cortex`? Wouldn't `loki` make more sense? ;-)

File: pkg/storage/chunk/client/azure/blob_storage_client.go
Comment (chaudum) [inline pkg/storage/chunk/client/azure/blob_storage_client.go:105]: Shouldn't the default be an empty string?
```suggestion
	f.StringVar(&c.UserAssignedID, prefix+"azure.user-assigned-id", "", "User assigned identity ID to authenticate to the Azure storage account.")
```

File: pkg/storage/chunk/client/azure/blob_storage_client.go
Comment (chaudum) [inline pkg/storage/chunk/client/azure/blob_storage_client.go:370]: There's no need for the empty string check:

```suggestion
	msiConfig := auth.MSIConfig{
		Resource: resource,
		ClientID: b.cfg.UserAssignedID,
	}
```

File: pkg/storage/bucket/azure/config.go
Comment (chaudum) [inline pkg/storage/bucket/azure/config.go:31]: Changing the default may break a user's setup if the container name is not explicitly set. Would an empty container name even be valid? :thinking:

**Review by chaudum — COMMENTED** [review]
This PR looks good, thanks for the contribution! I added a few minor comments.

File: pkg/storage/bucket/azure/config_test.go
Comment (stevehipwell) [reply pkg/storage/bucket/azure/config_test.go:16]: This is the default that was already set in the code where the value was actually being used, I've just made sure it's consistently set so this test needed changing.

File: pkg/storage/chunk/client/azure/blob_storage_client.go
Comment (stevehipwell) [reply pkg/storage/chunk/client/azure/blob_storage_client.go:105]: This looks like the default of the one I copied, I'll fix it.

File: pkg/storage/chunk/client/azure/blob_storage_client.go
Comment (stevehipwell) [reply pkg/storage/chunk/client/azure/blob_storage_client.go:370]: I agree that it's not actually needed, but I took the code from Thanos and appreciated that it made the flow easier to understand. I'm happy to remove it but just wanted to raise the readability question.

File: pkg/storage/bucket/azure/config.go
Comment (stevehipwell) [reply pkg/storage/bucket/azure/config.go:31]: I'm pretty sure that this value isn't being used from here as it contradicts the default being used elsewhere and I can't see an empty string working.

File: pkg/storage/bucket/azure/config_test.go
Comment (stevehipwell) [reply pkg/storage/bucket/azure/config_test.go:16]: I am interested as to why this is `cortex` but based on this and some of the comments I assume it was left over when the code was taken from Cortex?

Comment (chaudum) [reply pkg/storage/bucket/azure/config_test.go:16]: Yes, it's very likely a leftover.

**Review by chaudum — APPROVED** [review]
LGTM. Thanks again for the contribution! I don't think a changelog entry is necessary for this small change, is it @slim-bean ?

**Comment by stevehipwell** [comment]
@chaudum I wasn't sure if your CHANGELOG was auto generated or not, if it's manual I'm happy to add something as this is likely a useful change and an aid to adoption in Azure.

File: pkg/storage/bucket/azure/config_test.go
Comment (stevehipwell) [reply pkg/storage/bucket/azure/config_test.go:16]: I guess the bigger question is why is this value added to the bucket config when it's only used in the chunk config?

**Comment by stevehipwell** [comment]
@chaudum are you waiting on anything from me?

**Comment by cyriltovena** [comment]
Let's just merge conflict.

**Comment by stevehipwell** [comment]
@cyriltovena I need to wait for a reply to https://github.com/grafana/loki/pull/6015#issuecomment-1116171952 before I can rebase this branch but as soon as i hear back I'll get it back into a mergeable state.

**Comment by stevehipwell** [comment]
@cyriltovena I've rebased assuming that the changes made by @trevorwhitney in #6015 are correct even though I still haven't heard back.

CC @owen-d

File: docs/sources/configuration/_index.md
Comment (KMiller-Grafana) [inline docs/sources/configuration/_index.md:786]: Two small docs changes.

```suggestion
# This container must be created before running Cortex.
# CLI flag: -<prefix>.azure.container-name
[container_name: <string> | default = "loki"]

# Azure storage endpoint suffix without schema. The storage account name will
# be prefixed to this value to create the FQDN.
# CLI flag: -<prefix>.azure.endpoint-suffix
[endpoint_suffix: <string> | default = ""]

# Use Managed Identity to authenticate to the Azure storage account.
# CLI flag: -<prefix>.azure.use-managed-identity
[use_managed_identity: <boolean> | default = false]

# User-assigned identity ID to authenticate to the Azure storage account.
```

File: docs/sources/storage/_index.md
Comment (KMiller-Grafana) [inline docs/sources/storage/_index.md:303]:
```suggestion
    # Your Azure storage account name
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]
Small changes to capitalization needed.

File: docs/sources/configuration/_index.md
Comment (stevehipwell) [reply docs/sources/configuration/_index.md]: I swapped the text here to `Loki` based on prior work done.

**Comment by stevehipwell** [comment]
@cyriltovena how are we looking for getting this merged?

**Comment by stevehipwell** [comment]
@KMiller-Grafana could you update your approval or let me know if there is anything else required?

**Review by KMiller-Grafana — APPROVED** [review]
Thanks for the docs changes.

**Comment by stevehipwell** [comment]
@cyriltovena is there anything else needed before we can get this merged?

**Comment by chaudum** [comment]
Hey @stevehipwell sorry for letting you wait. Unfortunately, I don't have the permissions to merge, but I'll escalate it :)

**Review by kavirajk — APPROVED** [review]

**Comment by stevehipwell** [comment]
@chaudum is there an ETA for when these changes will be released? Also is there a pre-release version which would have these changes in it?

**Comment by stevehipwell** [comment]
@slim-bean it looks like v2.6.0 has this feature in it but it doesn't appear in the CHANGELOG?

**Pull request merged**
